docs/spec_draft.md

```diff
@@ -1267,6 +1268,70 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `body` function to `inputs` and `init_values` along the `dimensions`
+and produces a `result` tensor.
+
+The `body` function has to be associative, and it produces an
+implementation-defined behavior otherwise. Similarly, if `init_values` do not
+form identity under the `body` function, it produces an implementation-defined
+behavior.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | variadic number of scalars of any supported type |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `body`        | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `inputs`, `init_values` have the same element type.
+  * (C2) All `inputs` have the same shape.
+  * (C3) type(`inputs[k]`) $=$ type(`init_values[k]`) for all `k` $\in$ [0, N).
```

Comment:
```
* (C3) element_type(`inputs[k]`) $=$ element_type(`init_values[k]`) $=$
element_type(`results[k]`) for all `k` $\in$ [0, N).
```

Feedback:
  1. Only element types are equal, types are not never equal because inputs cannot be scalars.
  2. We also need to call out element types of results because they are never called out otherwise.